soccer/planning/SettlePathPlanner.cpp

```diff
@@ -0,0 +1,538 @@
+                // not when partial path is invalid
+                bool invalidPath = (partialPath && (path->start().motion.vel - partialPath->end().motion.vel).mag() > 0.01) ||
+                                    !partialPath;
+
+                // if (valid path) OR (3 invalid paths in a row)
+                // Use the current path and average in
+                //
+                // Increments the number of invalid path's if it's not valid
+                // If we hit that limit, we should change the path
+                // Always reset the number of invalid paths when a new value is averaged in
+                if (invalidPath) {
+                    numInvalidPaths++;
+                }
+
+                if (!invalidPath || numInvalidPaths > *_maxNumInvalidPaths) {
+                    interceptTarget = applyLowPassFilter<Point>(interceptTarget,
+                                                                targetRobotIntersection.pos,
+                                                                *_targetPointGain);
+
+                    averagePathTime = applyLowPassFilter<RJ::Seconds>(averagePathTime, t, *_targetPointGain);
+
+                    // Debug print
+                    // TODO: Remove
+                    if (numInvalidPaths > *_maxNumInvalidPaths) {
+                        std::cout << "Changing paths" << std::endl;
+                    }
+
+                    numInvalidPaths = 0;
+                }
+            }
+
+            foundInterceptPath = true;
+
+            break;
+        }
+    }
+
+    // Could not find a valid path that reach the point first
+    // Just go for the farthest point and recalc next time
+    if (!foundInterceptPath) {
+        if (!firstTargetPointFound) {
+            interceptTarget = ballVelIntercept;
+        } else {
+            interceptTarget = applyLowPassFilter<Point>(interceptTarget, ballVelIntercept, *_targetPointGain);
+        }
+
+        std::cout << "giving up in front" << std::endl;
+    }
+
+    // Make sure targetRobotIntersection is inside the field
+    if (!Field_Dimensions::Current_Dimensions.FieldRect().containsPoint(interceptTarget)) {
+        // Move backwards down the ball vel line until we are inside
+        // Super inefficient, but easy to do right now
+        // TODO: Change this to something smart
+        int ctr = 0;
+
+        while (!Field_Dimensions::Current_Dimensions.FieldRect().containsPoint(interceptTarget) && ctr < 100) {
+            interceptTarget -= 0.1*averageBallVel;
+            ctr++;
+        }
+
+        ctr = 0;
+        if (!Field_Dimensions::Current_Dimensions.FieldRect().containsPoint(interceptTarget)) {
+            interceptTarget += 10*averageBallVel;
+        }
+
+        while (!Field_Dimensions::Current_Dimensions.FieldRect().containsPoint(interceptTarget) && ctr < 100) {
+            interceptTarget += 0.1*averageBallVel;
+            ctr++;
+        }
```

Comment:
If someone has a clean way to do this, lmk. I'm having trouble finding a good search term in the couple minutes that I looked. Basically I want to limit the target point from moving outside the field. It the target point is outside, I want to select the intersection point of the ball velocity line with the field rectangle.

I can do the line rectangle intersection and pick the intersect point if it's closer, but I feel like there's a more direct way of doing this that we can throw into the geometry2d class.

Reply:
Huh? Why isn't a line-rect intersection what you want here? I guess technically you want to intersect with a ray, but I'm confused as to why that's still less "clean" than this.

Reply:
It's not that it's less clean, but that I feel like there is a more direct calculation that can be made to do this. If I do the intersection, it'll be the rect line intersection, a dot product to get the one in front of the moving ball, then a comparison to get the closest.

This class of problems has a lot of work out there due to the whole game design and graphics industry that I figured there might be a more unique and direct approach since I can see this use case in a variety of locations

Reply:
If we get a more direct approach, then we could just throw this into the rect class itself

Reply:
I don't think there is one, but a line segment-rectangle intersection is pretty fast anyway (our code for that allocates a vector for some reason, but I'm pretty sure it's still going to be faster than what you have here for most inputs and should only return two points: the robot's origin and the intersection point).

Reply:
I think we have a Line and and a Line segment class. You can immitate a velocity ray by making a long line segment and doing that intersection with the rectangle if it's implemented right.